### PR TITLE
python3Packages.gower: init at 0.1.2

### PR DESCRIPTION
**Pull request by b-rodrigues**
Link: https://pypi.org/project/gower/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by b-rodrigues** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.gower</li>
    <li>python311Packages.gower.dist</li>
    <li>python312Packages.gower</li>
    <li>python312Packages.gower.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/gower/default.nix
Comment (nim65s) [inline pkgs/development/python-modules/gower/default.nix:21]: Why do you need this ? 
Setting `pyproject = true;` instead doesn't work ?

**Comment by nim65s** [comment]
Can you add a `pythonImportsCheck` ?

File: pkgs/development/python-modules/gower/default.nix
Comment (b-rodrigues) [reply pkgs/development/python-modules/gower/default.nix:21]: Doing so results in the following error:

```
Creating a wheel...
* Getting build dependencies for wheel...
/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/dist.py:447: SetuptoolsDeprecationWarning: Invalid dash-separated options
!!

        ********************************************************************************
        Usage of dash-separated 'description-file' will not be supported in future
        versions. Please use the underscore name 'description_file' instead.

        This deprecation is overdue, please update your project and remove deprecated
        calls to avoid build errors in the future.

        See https://setuptools.pypa.io/en/latest/userguide/declarative_config.html for details.
        ********************************************************************************

!!
  opt = self.warn_dash_deprecation(opt, section)
/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py:72: _MissingDynamic: `license` defined outside of `pyproject.toml` is ignored.
!!

        ********************************************************************************
        The following seems to be defined outside of `pyproject.toml`:

        `license = 'MIT'`

        According to the spec (see the link below), however, setuptools CANNOT
        consider this value unless `license` is listed as `dynamic`.

        https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-project-metadata-the-project-table

        To prevent this problem, you can list `license` under `dynamic` or alternatively
        remove the `[project]` table from your file and rely entirely on other means of
        configuration.
        ********************************************************************************

!!
  _handle_missing_dynamic(dist, project_table)
/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py:72: _MissingDynamic: `dependencies` defined outside of `pyproject.toml` is ignored.
!!

        ********************************************************************************
        The following seems to be defined outside of `pyproject.toml`:

        `dependencies = ['numpy', 'scipy']`

        According to the spec (see the link below), however, setuptools CANNOT
        consider this value unless `dependencies` is listed as `dynamic`.

        https://packaging.python.org/en/latest/specifications/pyproject-toml/#declaring-project-metadata-the-project-table

        To prevent this problem, you can list `dependencies` under `dynamic` or alternatively
        remove the `[project]` table from your file and rely entirely on other means of
        configuration.
        ********************************************************************************

!!
  _handle_missing_dynamic(dist, project_table)
Traceback (most recent call last):
  File "/nix/store/mwvam3rmj6k3705z55griqzchlkjdajb-python3.12-pyproject-hooks-1.0.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
    main()
  File "/nix/store/mwvam3rmj6k3705z55griqzchlkjdajb-python3.12-pyproject-hooks-1.0.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 335, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/mwvam3rmj6k3705z55griqzchlkjdajb-python3.12-pyproject-hooks-1.0.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
    return hook(config_settings)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
    return self._get_build_requires(config_settings, requirements=[])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
    self.run_setup()
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/build_meta.py", line 313, in run_setup
    exec(code, locals())
  File "<string>", line 8, in <module>
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/__init__.py", line 108, in setup
    return distutils.core.setup(**attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 158, in setup
    dist.parse_config_files()
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/dist.py", line 603, in parse_config_files
    pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/pyprojecttoml.py", line 71, in apply_configuration
    return _apply(dist, config, filepath)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py", line 53, in apply
    _apply_project_table(dist, config, root_dir)
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py", line 79, in _apply_project_table
    corresp(dist, value, root_dir)
  File "/nix/store/l6vlavpj6wzcihb1qmrh1xh8wh22cmfb-python3.12-setuptools-72.1.0/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py", line 181, in _license
    _set_config(dist, "license", val["text"])
                                 ~~~^^^^^^^^
KeyError: 'text'


```

**Comment by b-rodrigues** [comment]
> Can you add a `pythonImportsCheck` ?

done

**Review by Kupac — APPROVED** [review]
Neat, the example from the package website works for me on x86_64-linux.

**Comment by b-rodrigues** [comment]
@natsukium is the PR ok ?

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gower/default.nix:20]: Please do not overwrite `patchPhase` because it will prevent us from applying patches.

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gower/default.nix:31]: Could you enable tests with `pytestCheckHook`?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook-using-pytestcheckhook

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gower/default.nix:33]:
```suggestion
    description = "Gower's distance calculation in Python";
```
> Not end with a period (or any punctuation for that matter).

https://nixos.org/manual/nixpkgs/unstable/#var-meta-description

**Review by natsukium — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/gower/default.nix
Comment (b-rodrigues) [reply pkgs/development/python-modules/gower/default.nix:13]: Opened the PR upstream: https://github.com/b-rodrigues/gower/pull/1

should we wait for it to be merged, or should I include this patch in this PR and then remove it once it's merged?

File: pkgs/development/python-modules/gower/default.nix
Comment (b-rodrigues) [reply pkgs/development/python-modules/gower/default.nix:20]: I have PR'ed the patch you suggested, by supposing we had to something like this, which phase should I have used instead? postPatch?

Comment (natsukium) [reply pkgs/development/python-modules/gower/default.nix:20]: Yes, we prefer to use `postPatch` instead.

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/gower/default.nix:13]: Thanks! We could patch without waiting for the merge and remove it when the version is updated.

File: pkgs/development/python-modules/gower/default.nix
Comment (nim65s) [reply pkgs/development/python-modules/gower/default.nix:33]: it does have unit tests: https://github.com/wwwjk366/gower/blob/master/tests/matrix_test.py
I don't know why pytestCheckHook does not find those

Comment (b-rodrigues) [reply pkgs/development/python-modules/gower/default.nix:33]: @natsukium should we skip the `pytestCheckhook`?

Comment (b-rodrigues) [reply pkgs/development/python-modules/gower/default.nix:34]: You were right, the PyPi tarball didn’t include the tests, but the code on github did!

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gower/default.nix:18]:
```suggestion
```
bad rebase

File: pkgs/development/python-modules/gower/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gower/default.nix:30]:
```suggestion
  build-system = [ setuptools ];

  dependencies = [
```

**Review by natsukium — COMMENTED** [review]
Besides this, lgtm.

**Review by natsukium — APPROVED** [review]
LGTM, thanks.

**Comment by natsukium** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 350747`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python311Packages.gower</li>
    <li>python312Packages.gower</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python311Packages.gower</li>
    <li>python312Packages.gower</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python311Packages.gower</li>
    <li>python312Packages.gower</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python311Packages.gower</li>
    <li>python312Packages.gower</li>
  </ul>
</details>

**Pull request merged**
